a11y-dialog.js

```diff
@@ -78,6 +78,7 @@
       }
     } else {
       this.container.setAttribute('data-a11y-dialog-native', '');
+      this.container.removeAttribute('aria-hidden');
```

Comment:
Pardon the silly question, but I’m not sure I get why we add this line?

Reply:
based on the example structure, suggested styles, and script, on initial load the modal is visible. it's only after the a11yDialog is instantiated that the script kicks in to determine to either add the `aria-hidden` or to mark it as `data-a11y-dialog-native`, actually hiding the modal. if there's lots of other JS in the page that needs to execute first, this can lead to what i'd call a "flash of unhidden modal" on load - the modal and the overlay are visible for a fraction of a second, before the script actually executes.

with this, i can already pre-set the dialog container as `aria-hidden` to begin with (so it's already visually hidden even before the script kicks in, thanks to the CSS).

now, if the browser does support `<dialog>` though, i want to remove this initially set `aria-hidden`, as otherwise a11yDialog won't remove it again (since it now relies on the native dialog handling).

Reply:
i've been seeing the "flash of unhidden modal" (or, in cases of browsers that support `<dialog>`, just the semi-transparent overlay) on a site i've been working on that has *a lot* of JS executing on page load, and in those cases it is very noticeable. but to give a more immediate idea, here's a screenshot of just using the live demo here with a breakpoint just before the script is run... which of course shows the overlay

> ![a11ydialog-stopped](https://user-images.githubusercontent.com/895831/95517855-c003b400-09b9-11eb-90d5-6d343c1cd628.png)

so, to counteract this, i added `aria-hidden` directly in the markup to the html itself. see how even before a11yDialog is instantiated, the modal is hidden (the much lighter grey overlay here in the screenshot is from devtools being paused, not from the overlay itself)

```
<div class="dialog" id="my-accessible-dialog" aria-hidden="true">
...
</div>
```

> ![modified stopped](https://user-images.githubusercontent.com/895831/95518649-68fede80-09bb-11eb-9221-2ce35b6c05f9.png)

in browsers that don't natively support `<dialog>`, this is fine as is and the modal still works (see for instance IE11 which has no problems here, even when there was already an initial `aria-hidden` added to the dialog, since in those browsers, the `aria-hidden` would have been set/removed anyway as part of its fallback logic to deal with browsers that don't support `<dialog>` natively.

> ![modal pops up in IE11](https://user-images.githubusercontent.com/895831/95518791-baa76900-09bb-11eb-9f97-a98d317d338a.png)

however, in browsers that *do* support `<dialog>` (e.g. Chrome with experimental web features enabled), that initial `aria-hidden` is never removed ... meaning that the modal never actually shows up (but you can see that the modal code itself does execute - see the console) because the styling that `display:none`s the modal when it has `[aria-hidden="true"]` still applies.

> ![in chrome, modal doesn't appear as aria-hidden is never removed](https://user-images.githubusercontent.com/895831/95518881-e62a5380-09bb-11eb-8d2f-48ba3867e5f2.png)

long story short: this one little `removeAttribute` as part of the initialisation fixes this latter case, leading to a best of both worlds:

- there's never going to be any "flash of unhidden modal" on page load (since even before the scripting kicks in, the modal is hidden by virtue of having `aria-hidden` directly in markup)
- browsers that don't support `<dialog>` behave just as they did before
- browsers that do support `<dialog>` effectively swap out the `aria-hidden="true"` for the `data-a11y-dialog-native` custom attribute on the dialog, still keeping it visually hidden. but now, the dialog then also works properly again

Reply:
That’s an outstanding explanation, thank you so much for taking the time. I recommend we add a code comment before this line with a link to your explanation (https://github.com/edenspiekermann/a11y-dialog/pull/117/files#r502043145). What do you think?